[master] Bug 573361: Add support for parameters within CriteriaBuilder case/coalesce expressions

dazey3 opened the pull request:
for #1150

Signed-off-by: Will Dazey <dazeydev.3@gmail.com>

lukasj commented on jpa/eclipselink.jpa.test.jse/src/it/java/org/eclipse/persistence/jpa/test/query/TestQueryCase.java:
```diff
@@ -328,8 +572,125 @@ public void testQueryCase3() {
         }
     }
 
+//    @Test
```
is this intentional?

dazey3 replied:
Yes, these were intentional. I commented out these 2 tests to disable them because they fail for an unrelated failure. I wrote the tests, and I believe they are valid tests, but they currently fail. I chose to only comment them out because I did not want to lose the test. I think using @Ignore is a much better idea though (I forgot that exists) and I will update to use that here instead.

lukasj commented on jpa/eclipselink.jpa.test.jse/src/it/java/org/eclipse/persistence/jpa/test/query/TestQueryCase.java:
```diff
@@ -238,7 +359,130 @@ public void testQueryCase2() {
+        }
+    }
+
+//    @Test
```
is this intentional?

lukasj reviewed (commented):
Too much of commented out code in tests for my taste. Would it make sense to have it in test with @Ignore or non-test method instead?

dazey3 commented:
@lukasj 
All of the commented out code in these tests is because either the test fails (`//@Test`) or the current JPA API doesn't even support what I feel should be supported (https://github.com/eclipse-ee4j/jpa-api/issues/315).
I chose to comment them out, instead of removing them entirely, because I didn't want the test to be lost or forgotten. All the tests I added are just taking the Literal test version and changing it to use Parameter markers. I feel all the tests are valid, but some either fail for reasons outside the scope of #1150 or are currently unsupported API calls

If you'd like to know more about the test failure, it has something to do with Constructor usage in CriteriaBuilder and type. The tests are failing with an error that EclipseLink cannot find the `org.eclipse.persistence.jpa.test.query.model.Dto01(String.class, Object.class, Object.class, Object.class)` constructor. I think this bug is more related to Constructor than CASE/COALESCE. I just didnt want to lose the test. 

I am happy to change to use @Ignore on `testQueryCaseLiterals3` and `testQueryCaseParameters3`. As to the `// Currently unsupported by the JPA API` commented out code. Would you rather I just delete it? I can be fine with that if you don't want it.

lukasj commented:
> @lukasj
> All of the commented out code in these tests is because either the test fails (`//@Test`) or the current JPA API doesn't even support what I feel should be supported ([eclipse-ee4j/jpa-api#315](https://github.com/eclipse-ee4j/jpa-api/issues/315)).
> I chose to comment them out, instead of removing them entirely, because I didn't want the test to be lost or forgotten.

I'm fine with keeping it there as long as there is an explanation and/or a pointer to existing issue (if it exists). Imagine someone with limited knowledge comes through it - such person may just delete comments simply because "it was useless comment"...

dazey3 commented:
@lukasj
> Imagine someone with limited knowledge comes through it - such person may just delete comments simply because "it was useless comment"...
That is a fair statement. I will add a comment to help explain better the reason for the commented out code. 
Also, are you ok with adding references to issues in GH? For instance, referencing https://github.com/eclipse-ee4j/jpa-api/issues/315?

dazey3 commented:
@lukasj 
Ok, I updated the test to be perhaps a bit cleaner/clearer. I apologize for using a git force push, but I get errors using a standard git push:
```
    ! [rejected]              573361_master -> 573361_master (non-fast-forward)
    error: failed to push some refs to 'git@github.com:dazey3/eclipselink.git'
    hint: Updates were rejected because the tip of your current branch is behind
    hint: its remote counterpart. Integrate the remote changes (e.g.
    hint: 'git pull ...') before pushing again.
```
Everything is up to date and rebased. No idea why it does this! I just know that force has always worked.

lukasj commented:
`git rebase master && git push -f` is my workflow to keep my own commits "untouched" and on top

lukasj reviewed (approved):
LGTM

dazey3 commented:
I missed an @Ignore on one of the tests and it failed in travis. I fixed it and pushed an update that just contained the new @Ignore
